### ISSUE TITLE
perf: reconsider crypto/rand use in UUID allocation for txn IDs

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#30236]
We currently use V4 UUIDs as transaction IDs. Our UUID library, `github.com/satori/go.uuid`, uses a cryptographically secure random number generator (`crypto/rand`) to [populate](https://github.com/satori/go.uuid/blob/f58768cc1a7a7e77a3bd49e98cdd21419399b6a3/generator.go#L200) the UUID. In #30233, we observed that this was responsible for **77%** of mutex contention during a `kv95 --concurrency=64` run on my laptop:

<img width="1194" alt="screen shot 2018-09-14 at 1 18 20 pm" src="https://user-images.githubusercontent.com/5438456/45565399-761f6280-b821-11e8-9fca-6bd874a2dfe7.png">

I ran an experiment where I swapped this over to a cryptographically insecure `math/rand` rng (which still uses a global lock) and saw this lock contention completely disappear. I also saw throughput jump by about **2.1%**.

<img width="1174" alt="screen shot 2018-09-14 at 1 18 29 pm" src="https://user-images.githubusercontent.com/5438456/45565402-79b2e980-b821-11e8-9705-8804315d718d.png">

I don't think we need a cryptographically secure random byte generator for our transaction IDs, so we should consider making this change for real. `github.com/satori/go.uuid` won't let us swap out the rng, but `github.com/google/uuid` will, and we already have an open issue to replace our uuid library: 
https://github.com/cockroachdb/cockroach/issues/26332.

Also, FWIW in the long run, we might be replacing UUID txn ids entirely. See https://github.com/cockroachdb/cockroach/pull/24194.

@m-schneider this got me thinking about https://github.com/cockroachdb/cockroach/issues/26178. If/once https://github.com/cockroachdb/cockroach/pull/30233 is merged, we should take a look at mutex profiles before and after the throughput cliff we're observing in that issue.

**Comment by jordanlewis** [comment]
Yes!! Really happy to see this - I've noticed UUID generation taking a lot of time in profiles but I didn't think we could fix it.

**Comment by petermattis** [comment]
I seem to recall looking at this in the past and discovering that crypto generation on Mac OS X was super slow in comparison to Linux. You might want to verify the perf increase on a Linux box.

**Comment by nvanbenschoten** [comment]
I looked into this at the time and confirmed that `uuid.NewV4` did not have the same impact on mutex profiles on Linux machines as it does Mac. 

However, I'm now seeing it come up on a single node `m5d.12xlarge` (48 vCPU) cluster in CPU profiles. It's responsible for a little under 2% of the total CPU utilization. All of the time is spent in `internal/syscall/unix.GetRandom`'s `getrandom(2)` syscall. I don't know anything about the `/dev/urandom` device's implementation, but posts like http://drsnyder.us/2014/04/16/linux-dev-urandom-and-concurrency.html and https://news.ycombinator.com/item?id=7610273 paint a clear enough picture that it's going to be a concurrency bottleneck, especially on larger machines. Since it doesn't sound like there's any reason why we need a cryptographically secure random byte generator for these IDs, I think we should make the change.
